product_end_of_life/models/product.py

```diff
@@ -0,0 +1,187 @@
+            ('eol_date', '<=', end_date),
+            ('type', '=', 'product')])
+
+        # send email to recipients from the Product Eol Discussion Channel.
+        self.email_product_eol_list(product_eol_ids, run_day)
+        return True
+
+
+class ProductVariants(models.Model):
+    _inherit = 'product.product'
+
+    eol_date = fields.Date(
+        string='End of Life Date'
+    )
+
+    @api.model
+    def name_search(self, name, args=None, operator='ilike', limit=100):
+        result = super(ProductVariants, self).name_search(
+            name=name,
+            args=args,
+            operator=operator,
+            limit=limit)
+        context = self._context or {}
+        update_result = []
+        today = fields.Date.today()
+        if 'source_from' in context and \
+                context.get('source_from') == 'purchase_order':
+            if result:
+                for res in result:
+                    for product in self.browse(res[0]):
+                        # Exclude expired products
+                        if product.eol_date and product.eol_date <= today:
+                            continue
+                        else:
+                            # Design product code, name
+                            def _name_get(d):
+                                name = d.get('name', '')
+                                code = \
+                                    context.get('display_default_code', True) \
+                                    and d.get('default_code', False) or False
+                                if code:
+                                    name = '[%s] %s' % (code, name)
+                                return (d['id'], name)
+                                # prepare product attributes
+                            variant = ", ".join([v.name for v in
+                                                 product.attribute_value_ids])
+                            name = variant and "%s (%s)" % \
+                                (product.name, variant) or product.name
+                            mydict = {
+                                'id': product.id,
+                                'name': name,
+                                'default_code': product.default_code,
+                            }
+                            update_result.append(_name_get(mydict))
+        else:
+            update_result = result
+        return update_result
+
+    @api.model
+    def search_read(self, domain=None, fields=None, offset=0, limit=None,
+                    order=None):
+        res = super(ProductVariants, self).search_read(domain=domain,
+                                                       fields=fields,
+                                                       offset=offset,
+                                                       limit=limit,
+                                                       order=order)
+        today = str(datetime.today().date())
```

Comment:
in this function when fields is used to get today's date an error 'AttributeError: 'list' object has no attribute 'Date'' since as you can see on the function parameter that field is being used.